### ISSUE TITLE
Add support for "Sign In with Apple"

### ISSUE THREAD

**Issue by smokris** [hybridauth/hybridauth#1095]
Would the Hybridauth maintainers be open to including support for the [Sign In with Apple](https://developer.apple.com/sign-in-with-apple/) service?

Additional info: https://developer.apple.com/documentation/signinwithapplejs/configuring_your_webpage_for_sign_in_with_apple

**Comment by ApacheEx** [comment]
I agree that would be great.

All needed info can be found here: https://developer.apple.com/documentation/signinwithapplejs/incorporating_sign_in_with_apple_into_other_platforms

Seems, it's not a big deal to do it.

**Comment by ashenwolf** [comment]
Just wondering if anyone already managed to implement Apple sign-in as hybridauth provider?

**Comment by hhoweson** [comment]
Any progress on this? I would like to implement hybridauth on my site but require apple login to be available ):

**Comment by ApacheEx** [comment]
this should be quite easy to implement. If no one else wants to do it I will take a look on it this weekend.

**Comment by hhoweson** [comment]
That would be great! @ApacheEx

**Comment by hhoweson** [comment]
@ApacheEx Any news on this? I really appreciate you taking this on and don't mean to rush you (:

**Comment by hhoweson** [comment]
@ApacheEx Any news on this? I really appreciate you taking this on and don't mean to rush you (:

**Comment by ApacheEx** [comment]
sorry guys, recently I've checked this but struggle with 2 things:
- from the first look we need to have possibility to encode/decode JWT. I guess would be nice to avoid this 3rd party dependency.
- to check if it works properly it requires to have Apple Developer certificate but I don't have it 😭

**Comment by hhoweson** [comment]
@ApacheEx If you're able to get the rest of it working I would be happy to test it with my Apple Developer certificate

**Comment by sputnik73** [comment]
@ApacheEx thank you, I will try to do some testing with our certificate, too.

**Comment by hhoweson** [comment]
@ApacheEx Any updates? Let me know if there is anything I can do to help!

**Comment by sputnik73** [comment]
I managed to write my own small Apple provider-class meanwhile, at least authentication is already working. 

However right now I simply use a static secret created by client_secret.rb Script from https://developer.okta.com/blog/2019/06/04/what-the-heck-is-sign-in-with-apple

Unfortunately `getUserProfile()` will not be available since Apple only delivers _email, firstname, lastname_ a single time with the first authorization request.

I needed to add an _AuthorizeUrlParametersEncType_ property to The OAuth2 because Apple requires `PHP_QUERY_RFC3986` encryption in `http_build_query`.

Right now the code is work in progress, if interested I can provide it in some days. Has anybody a good solution for secret generation?

**Comment by gs999918** [comment]
Hi,
 
I've been following this thread for quite some time now.  I've been waiting for someone to contribute back to the project so we can start implementing into our own individual projects.

For those, of us that have mobile apps, Apple will start requiring Apple Sign In on your app if you in any way have used hybridauth in the past for your app.  I.e. you must use Apple-Sign-In if you use ANY OTHER SIGN IN SERVICE, such as Google, Twitter, or Facebook, Yahoo, Amazon, etc.

The deadline is about 1.5 months away.  If at some point someone can contribute a working library file back to the project to make Apple a provider under hybrid auth, that would be great.

**Comment by sputnik73** [comment]
> The deadline is about 1.5 months away. If at some point someone can contribute a working library file back to the project to make Apple a provider under hybrid auth, that would be great.

I will try (I am not involved in the project), since we need it for our apps, too. However as mentioned there will be no `getUserProfile()` since Apple simply does not provide an API for that.
 
As far as I see the app will have to deliver _email, firstname, lastname_ to the server if this information is required there. So for registration of new users, sending the token to the server will not not be sufficiant, since  this information is only delivered with the first authorization.

**Comment by sputnik73** [comment]
You can find the current state and [some documentation](https://github.com/sputnik73/hybridauth-sign-in-with-apple/blob/master/APPLE.md)  at https://github.com/sputnik73/hybridauth-sign-in-with-apple

The Apple provider class is currently nearly identical to the parent OAuth2 class and only sets a parameter and a different query string encoding. Finding all keys/ids and creating the secret was the harder thing. Creating the secret this way is only for testing since the it has only a lifetime of 6 months, it should be replaced by [something like this...](https://stackoverflow.com/questions/56459075/how-to-verify-code-from-sign-in-with-apple)

Any suggestions are welcome since the hybridauth code and contributing to projects in general are new to me.

**Comment by omidborjian** [comment]
I spent all day on Apple Sign In to get it to work. @sputnik73 thanks for pointing me in the right direction. I made a few tweaks which may help others. So for anyone that can't wait for the official Provider, here we go:

Assuming you are suing @sputnik73's Apple.php provider and the latest HybridAuth. 

In the configurations, the "key/id" and "secret" cannot be empty. Even though @sputnik73 code generates the secret, HybridAuth will throw and exception if the secret is empty. In his doc, it sets secret as "foo". don't remove it.

Secondly, I kept getting a "Signature verification invalid" when the validateAccessTokenExchange of the Apple provider gets executed. It's using Apple's keys to decode the JWT id_token so you can get the user. I simplified so you don't even need JWT or the JWKToPEM Converter anymore (for this step):

` protected function validateAccessTokenExchange($response)
    {
        $data = (new Data\Parser())->parse($response);

        $collection = new Data\Collection($data);

        if (!$collection->exists('id_token')) {
            throw new InvalidAccessTokenException(
                'Provider returned an invalid access_token: ' . htmlentities($response)
            );
        }

       //JWT splits the string to 3 components 1)first is header 2)is payload 3)is signature
        $payload = explode('.', $collection->get('id_token'))[1];
        $payload = json_decode(base64_decode($payload));

        $this->storeData('token_payload', $payload);

        parent::validateAccessTokenExchange($response);
    }
`    

That's it. Let me know if you see something I missed

**Comment by sputnik73** [comment]
Thank you for the feedback. I added a hint to the description that the a non empty secret string is required. 

Your solution to get the token payload from `validateAccessTokenExchange` is nice, because it is simple, without external dependencies and doesn't need to lookup the current public keys. I will look up what is best practice for this - or maybe provide a configurable parameter and let the users decide if they want to verify the response using the public key.

**Comment by sputnik73** [comment]
Updates/fixes:

- missing namespace alias fixed
- token extraction without signature validation implemented (@omidborjian thank you!). Users can disable validation with `"verifyTokenSignature" => false`
- if token validation is enabled, `composer require codercat/jwk-to-pem` is required (documentation fixed/added)

I can confirm something is currently broken with signature validation, I will try to fix it tomorrow.

**Comment by endisha** [comment]
Hello guys this is Apple's provider. May help you

```
<?php
use Hybridauth\Adapter\OAuth2;
use Hybridauth\Data;
use Hybridauth\Exception\UnexpectedApiResponseException;
use Hybridauth\Exception\InvalidApplicationCredentialsException;
use Hybridauth\User;

/**
 * Apple OAuth2 provider adapter.
 */
class Apple extends OAuth2
{
    /**
     * {@inheritdoc}
     */
    public $scope = 'name email';

    /**
     * {@inheritdoc}
     */
    protected $apiBaseUrl = 'https://appleid.apple.com/auth/';

    /**
     * {@inheritdoc}
     */
    protected $authorizeUrl = 'https://appleid.apple.com/auth/authorize';

    /**
     * {@inheritdoc}
     */
    protected $accessTokenUrl = 'https://appleid.apple.com/auth/token';

    /**
     * {@inheritdoc}
     */
    protected $apiDocumentation = 'https://developer.apple.com/documentation/sign_in_with_apple';

    /**
    * {@inheritdoc}
    */
    protected function initialize(){
        parent::initialize();
        $this->AuthorizeUrlParameters['response_mode'] = 'form_post';
    }


    /**
     * Decodes a JWT string into a PHP object
     * 
     * @param  Data\Collection  $keys
     * @param  int $key
     * @return array
     */
    protected function decodeToken($keys, $key=0){

        $parsedKeyData = $keys->filter('keys')->get($key);
        $parsedPublicKey= \Firebase\JWT\JWK::parseKey($parsedKeyData);
        $publicKeyDetails = openssl_pkey_get_details($parsedPublicKey);

        if(!isset($publicKeyDetails['key'])) {
            throw new \Exception('Invalid public key details.');
        }

        return \Firebase\JWT\JWT::decode($this->getStoredData('id_token'), $publicKeyDetails['key'], array($parsedKeyData->alg));
    }



    /**
     * {@inheritdoc}
     */
    public function getAccessToken()
    {
        $tokenNames = [
            'access_token',
            'id_token',
            'access_token_secret',
            'token_type',
            'refresh_token',
            'expires_in',
            'expires_at',
        ];

        $tokens = [];

        foreach ($tokenNames as $name) {
            if ($this->getStoredData($name)) {
                $tokens[ $name ] = $this->getStoredData($name);
            }
        }

        return $tokens;
    }

    /**
     * {@inheritdoc}
     */
    protected function validateAccessTokenExchange($response)
    {
        $collection = parent::validateAccessTokenExchange($response);

        $this->storeData('id_token', $collection->get('id_token'));

        return $collection;
    }

    /**
     * {@inheritdoc}
     */
    public function getUserProfile()
    {

        $response = $this->apiRequest('keys');

        $keys = new Data\Collection($response);

        if (! $keys->exists('keys')) {
            throw new UnexpectedApiResponseException('Provider API returned an unexpected response.');
        }

        $key = 0;
        $error = [];
        do{
            try {
                $payload = $this->decodeToken($keys, $key);
                $tryAgain = false;
                $error = false;
            }catch(\Exception $e) {
                $tryAgain = $key < $keys->count();
                if($e instanceof \Firebase\JWT\ExpiredException){
                    $tryAgain = false;
                }
                $error = $e->getMessage();
                $key++;
            }
        } while($tryAgain);

        if($error){
            throw new \Exception($error);
        }
        
        
        $data = new Data\Collection($payload);

        $userProfile = new User\Profile();
        $userProfile->email = $data->get('email');

        return $userProfile;
    }

}
```


To generate client secret use this package Lcobucci\JWT to create use ES256 

```
use Lcobucci\JWT\Builder;
use Lcobucci\JWT\Signer\Key;
use Lcobucci\JWT\Signer\Ecdsa\Sha256;

$privateKey = file_get_contents(__DIR__.'/private_key.pem');
$clientId = '';
$teamId = '';
$keyId = '';
$alg = 'ES256';
$clientSecretToken = (string) (new Builder())
						->withHeader('kid', $keyId)
						->withHeader('type', 'JWT')
						->withHeader('alg', $alg)
						->issuedBy($teamId) 
						->issuedAt(time()) 
                        ->expiresAt(time() + 3600)
                        ->permittedFor('https://appleid.apple.com') 
                        ->relatedTo($clientId)
                        ->getToken(new Sha256(), new Key($privateKey)); 
```
Config:
```
"Apple" => [
	'callback' => \Hybridauth\HttpClient\Util::getCurrentUrl(),
	"enabled" => true,
	"keys"    => [
		'private_key' => 'PRIVATE_KEY',
		'id' => '', //service id
		'team_id' => '', //team id
		'key_id' => '', // key_id
		'secret' => 'CLIENT_SECERT',
	],
],
```

**Comment by sputnik73** [comment]
@endisha Thx, your token handling looks is much better. However  `\Firebase\JWT\JWK::parseKey(array $jwk)` is private at https://github.com/firebase/php-jwt/blob/master/src/JWK.php  -  which source did you use?

**Comment by sputnik73** [comment]
Updates to https://github.com/sputnik73/hybridauth

- prevent problems when there is a clock skew times between the signing and verifying servers by adding a leeway
- check if the access_token is expired in `isConnected()`
- test the signature with all provided public keys
- token handling moved to getUserProfile() 

With these changes signature verification and token handling should work more reliable now.

**Comment by endisha** [comment]
> @endisha Thx, your token handling looks much better. However `\Firebase\JWT\JWK::parseKey(array $jwk)` is private at https://github.com/firebase/php-jwt/blob/master/src/JWK.php - which source did you use?
@sputnik73
https://github.com/fproject/php-jwt > this old fproject/php-jwt package

**Comment by sputnik73** [comment]
@endisha thank you for your code, it es very helpful. How should we continue?

- Do you want to create a new fork or continue using https://github.com/sputnik73/hybridauth for a pull request later at https://github.com/hybridauth/hybridauth?
- Should we use fproject/php-jwt or firebase/php-jwt?
- In my version, secret generation is done in the provider class, which is more comfortable to use. Do you prefer to remove it from there?
- I suggest to assign identifier, firstname and lastname in `getUserProfile()`
- According to Apple blanks should be encoded using `%20` in `http_build_query` (PHP_QUERY_RFC3986)
- `\Firebase\JWT\JWT::$leeway = 60;` should be set to prevent timing errors

**Comment by sputnik73** [comment]
To use my fork in your composer configuration as a workaround until there is an Apple provider class in the official hybridauth/hybridauth available:

1) add these lines to `composer.json`:
```
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/sputnik73/hybridauth"
        }
    ],
```

2) replace the hybridauth version with the branch name:
`"hybridauth/hybridauth": "dev-master",` 

3) php composer.phar update `hybridauth/hybridauth`

**Comment by hhoweson** [comment]
Is there vanilla support for "Sign in with apple" yet? Looks like there have been a number of different methods worked on and some seem to have been pushed to live however the documentation has not yet been updated at https://hybridauth.github.io/providers.html

**Comment by sputnik73** [comment]
> Is there vanilla support for "Sign in with apple" yet? Looks like there have been a number of different methods worked on and some seem to have been pushed to live

There are pull requests  ready for discussion/review/merge for two implementations right now:

- https://github.com/hybridauth/hybridauth/pull/1175 from @endisha 
- https://github.com/hybridauth/hybridauth/pull/1178 from me (@sputnik73)

None of these apple Provider classes is "live" in the [current master branch of hybridauth](https://github.com/hybridauth/hybridauth/tree/master/src/Provider) yet.

For our use case therefore we use my hybridauth fork https://github.com/sputnik73/hybridauth until there is an official "Sign in with apple" Provider in hybridauth. It is in production use now since at least a month and without problems yet.

The implementation posted by @endisha on 15/05/2020 without discussion was surprising. I compared it to my existing code and adopted parts he solved better - and he did, thx & credits for more experienced token handling go to him. However my priority is to have an Apple provider class which is exactly as convenient to use as all other provider classes and "works", although Apple implemented OAuth2 in a more complicated way (eg. secrets).  Since I got no feedback, I continued working on my code and a (hopefully) useful documentation. Any suggestions/improvements welcome of course.

**Comment by hhoweson** [comment]
Thanks for filling me in @sputnik73, I will give your implementation a try and let you know how I get on! Out of interest, what is currently holding your implementation back from being pulled to master? Would be great not to have to worry about future updates!

**Comment by sputnik73** [comment]
> Thanks for filling me in @sputnik73, I will give your implementation a try and let you know how I get on!

Thanks! 👍 

>  Out of interest, what is currently holding your implementation back from being pulled to master? Would be great not to have to worry about future updates!

I guess the maintainers simply have lack of time, a review is time-consuming since the Apple API is not trivial. For my part, we will continue merging upstream changes and fix bugs in my fork because we rely on hybridauth ourselves. But Of course it would be more convenient to have it included in hybridauth.

**Comment by sputnik73** [comment]
@hhoweson good news: the maintainers just accepted and merged my Apple provider class :-)
https://github.com/hybridauth/hybridauth/pull/1178

**Comment by hhoweson** [comment]
@sputnik73 Haha, was just reading through it, great work! Thank you 😄

**Comment by ApacheEx** [comment]
yeah, lack of time - you are absolutely right. This is why contribution is always welcome.

Released as [3.4](https://github.com/hybridauth/hybridauth/releases/tag/3.4.0)
